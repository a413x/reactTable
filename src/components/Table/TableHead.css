.table th{
  position: relative;
  cursor: pointer;
  user-select: none;
  overflow: hidden;
  width: 16%;
}
.table th:first-child{
  width: 8%;
}
.table th:nth-child(0n+4){
  width: 36%;
}
.table th:last-child{
  width: 24%;
}
.asc:after,.desc:after{
  content: '';
  position: absolute;
  top: 20px;
  right: 0;
  border-width: 5px 5px 0;
  border-style: solid;
  border-color: #33a transparent;
}
.desc:after{
  border-width: 0 5px 5px;
}
@media (min-width: 576px) {
  .table th{
    overflow: visible;
  }
}
