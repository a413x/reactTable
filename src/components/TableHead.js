export const tableColumns = [
  {name: 'id', title: '#'},
  {name: 'firstName', title: 'First name'},
  {name: 'lastName', title: 'Last name'},
  {name: 'email', title: 'E-mail'},
  {name: 'phone', title: 'Phone'},
]

export const TableHead = () => {
  return (
    <thead>
      <tr>
        {tableColumns.map((colObj, ind) =>
          <th key = {'th-' + ind}>
            {colObj.title}
          </th>
        )}
      </tr>
    </thead>
  )
}
