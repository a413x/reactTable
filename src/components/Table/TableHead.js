import {useState} from 'react'
import './TableHead.css'

export const tableColumns = [
  {name: 'id', title: '#'},
  {name: 'firstName', title: 'First name'},
  {name: 'lastName', title: 'Last name'},
  {name: 'email', title: 'E-mail'},
  {name: 'phone', title: 'Phone'},
]

export const TableHead = ({sortCallback}) => {
  const [activeCol, setActiveCol] = useState({name: '', order: 'asc'})

  const onClick = (colObj) => {
    let order = activeCol.order
    if(activeCol.name === colObj.name){
      order = order === 'asc' ? 'desc' : 'asc'
    }else{
      order = 'asc'
    }
    setActiveCol({name: colObj.name, order: order})
    sortCallback(colObj, order)
  }

  return (
    <thead>
      <tr>
        {tableColumns.map((colObj, ind) =>
          <th
            key = {'th-' + ind}
            onClick = {() => onClick(colObj)}
            className = {
              activeCol.name === colObj.name ? activeCol.order : ''
            }
          >
            {colObj.title}
          </th>
        )}
      </tr>
    </thead>
  )
}
