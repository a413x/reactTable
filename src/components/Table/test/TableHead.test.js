import {render, fireEvent} from '@testing-library/react'
import {TableHead, tableColumns} from '../TableHead.js'

describe('TableHead component tests', () => {
  const mock_sortCallback = jest.fn()

  const renderWithTable = (sortedCol = {name: '', order: ''}) => {
    const table = document.createElement('table')
    return render(
      <TableHead
        sortCallback={mock_sortCallback}
        sortedCol = {sortedCol}
      />,
      { container: document.body.appendChild(table) }
    )
  }

  test('renders properly', () => {
    const {container} = renderWithTable()
    expect(container).toMatchSnapshot()
  })

  test('ascending sort arrow correctly displays', ()=>{
    const column = tableColumns[2]
    const {getByText} = renderWithTable({name: column.name, order: 'asc'})
    const th = getByText(column.title)
    expect(th.classList.contains('asc')).toBe(true)
  })

  test('descending sort arrow correctly displays', ()=>{
    const column = tableColumns[2]
    const {getByText} = renderWithTable({name: column.name, order: 'desc'})
    const th = getByText(column.title)
    expect(th.classList.contains('desc')).toBe(true)
  })

  test('sortCallback function calls on table header cell click', () => {
    const {getByText} = renderWithTable()
    const column = tableColumns[3]
    const th = getByText(column.title)
    fireEvent.click(th)
    expect(mock_sortCallback).toHaveBeenCalledWith(column, 'asc')
  })
})
