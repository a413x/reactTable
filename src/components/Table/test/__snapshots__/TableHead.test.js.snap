// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableHead component tests renders properly 1`] = `
<table>
  <thead>
    <tr>
      <th
        class=""
      >
        #
      </th>
      <th
        class=""
      >
        First name
      </th>
      <th
        class=""
      >
        Last name
      </th>
      <th
        class=""
      >
        E-mail
      </th>
      <th
        class=""
      >
        Phone
      </th>
    </tr>
  </thead>
</table>
`;
